### ISSUE TITLE
With dracut 100, nvme kernel drivers are not included in hostonly mode

### ISSUE THREAD

**Issue by dalto8** [dracut-ng/dracut-ng#130]
**Describe the bug**
When building the initrd with `dracut --hostonly --no-hostonly-cmdline` nvme kernel drivers are not included even though there are nvme devices installed and used in the system.  Downgrading to 059 includes these drivers. 

**Distribution used**
Arch Linux

**Dracut version**
100

**Init system**
systemd

Please let me know if I can provide any further information.

**Comment by dalto8** [comment]
Here is the log with dracut -v https://pastebin.mozilla.org/vbT0oBGO

I suspect it will work if I add the driver manually since it is added when it isn't in hostonly mode.

If needed, I can build and test v60.

**Comment by dalto8** [comment]
This is in the logs:
```
dracut[I]: *** Including module: kernel-modules ***
```

Is there something else I should check?

**Comment by dalto8** [comment]
> I see no change around the nvme driver (file changed but not the line) - https://github.com/dracut-ng/dracut-ng/blob/main/modules.d/90kernel-modules/module-setup.sh#L30

It isn't even reaching that line.  It isn't entering this `if` block: https://github.com/dracut-ng/dracut-ng/blob/main/modules.d/90kernel-modules/module-setup.sh#L107

**Comment by dalto8** [comment]
I built the Arch packages for 60-1 and 60-2.(The latter removes some patches).

Both those versions include the nvme driver.  Only when upgrading to 100 does it stop being included.

**Comment by freswa** [comment]
@dalto8 Could you post your dracut config and the command you run dracut with please?

**Comment by dalto8** [comment]
> @dalto8 Could you post your dracut config and the command you run dracut with please?

The command is `dracut --hostonly --no-hostonly-cmdline initrd 6.8.2-arch2-1`

The config file is unchanged but I do have a drop in that has this in it.

```
omit_dracutmodules+=" network cifs nfs nbd brltty i18n "
#hostonly_mode=strict
compress="zstd"
```
I commented out `hostonly_mode=strict` as my first troubleshooting step.

> Perhaps caused by this #36 ?

Let me revert that and test.

**Comment by dalto8** [comment]
> Perhaps caused by this #36 ?

Yes.  That seems be it.  Specifically, the change to [line 1670](https://github.com/dracutng/dracut-ng/blob/1cb7754d1278814786e6d98e4915cf59fb667cc5/dracut.sh#L1670).

If I revert that change, it works again.

**Comment by aafeijoo-suse** [comment]
> Yes. That seems be it. Specifically, the change to [line 1670](https://github.com/dracutng/dracut-ng/blob/1cb7754d1278814786e6d98e4915cf59fb667cc5/dracut.sh#L1670).

If `find_block_device` returns with an error, the following code is borked. E.g., when `/etc` is mounted as overlay, it's adding the empty string '' as host device, breaking functions that are expecting a real device, like `get_maj_min`.

So, as a first action it would be great to check what's the output of `dracut --debug --hostonly --no-hostonly-cmdline initrd 6.8.2-arch2-1` with the change added again.

**Comment by dalto8** [comment]
> If find_block_device returns with an error, the following code is borked. E.g., when /etc is mounted as overlay, it's adding the empty string '' as host device, breaking functions that are expecting a real device, like get_maj_min.

I suspect it may be hard to identify all the cases that could cause this condition to occur but what about scenarios in which `/` is not mounted directly on a block device?  In this case, the underlying device still needs to be added to the list.

The reason it fails in my case is probably that the code has cases to handle finding the underlying block devices for zfs datasets and btrfs filesystems.  However, if the mountpoint is mounted on something other than block device, that `continue` causes those blocks of code to be skipped and the underlying devices are never identified.

**Comment by aafeijoo-suse** [comment]
A debug log would help us to understand what is happening in your system, to add a proper fix. As I said, the revert also breaks another systems. You do not need to break you system, you can just modify the line, create `dracut --debug --hostonly --no-hostonly-cmdline test.img 2>&1 &> dracut.log`, and attach the `dracut.log` file.

**Comment by dalto8** [comment]
Yes, I already reviewed the logs and analyzed the relevant code to determine the issue.  That is where my description above came from.

My mountpoints are not mounted on block devices.  However, the prior fix skipped the entire remainder of the loop, not just adding that specific device.

I created https://github.com/dracut-ng/dracut-ng/pull/155 that works for my situation and hopefully resolves the issue you were facing as well.

**Comment by StormBytePP** [comment]
I am suffering this issue in version 105, was this reintroduced again?

The workaround is to add force_drivers+=" nvme " because despite the module nvme is present in the image it is not automatically modprobed for some reason and therefore the image is unbootable.

Happens both with and without hostonly
